src/pip/_internal/commands/list.py

```diff
@@ -160,32 +170,38 @@ def run(self, options, args):
             packages = self.get_uptodate(packages, options)
 
         self.output_package_listing(packages, options)
+        return SUCCESS
 
     def get_outdated(self, packages, options):
+        # type: (List[Distribution], Values) -> List[Distribution]
         return [
             dist for dist in self.iter_packages_latest_infos(packages, options)
             if dist.latest_version > dist.parsed_version
         ]
 
     def get_uptodate(self, packages, options):
+        # type: (List[Distribution], Values) -> List[Distribution]
         return [
             dist for dist in self.iter_packages_latest_infos(packages, options)
             if dist.latest_version == dist.parsed_version
         ]
 
     def get_not_required(self, packages, options):
-        dep_keys = set()
+        # type: (List[Distribution], Values) -> List[Distribution]
+        dep_keys = set()  # type: Set[Distribution]
         for dist in packages:
             dep_keys.update(requirement.key for requirement in dist.requires())
-        return {pkg for pkg in packages if pkg.key not in dep_keys}
+        return [pkg for pkg in packages if pkg.key not in dep_keys]
```

Comment:
Why this change?

Reply:
This function, and the two before that are being used as follows

```
        if options.not_required:
            packages = self.get_not_required(packages, options)

        if options.outdated:
            packages = self.get_outdated(packages, options)
        elif options.uptodate:
            packages = self.get_uptodate(packages, options)
```

and `get_outdated` and `get_uptodate` are returning lists, so I change the return type of `get_not_required` to `list` as well. 

I actually can return something like `list({pkg for pkg in packages if pkg.key not in dep_keys})` as well, perhaps that might be better?

Reply:
I have gone ahead and changed it to `list({pkg for pkg in packages if pkg.key not in dep_keys})`

Reply:
I'd suggest adding a comment, as it's not obvious on first reading why you need the {...} (to remove duplicates, presumably?) A more invasive, but probably better, change would be to make packages a set (which would ripple back up to `run()` where packages was first created.

Reply:
Hi @pfmoore 

For now I have added the comment describing why it is a set, and why do we cast it as a set.
Making packages as a set is actually a better idea, but I would probably create a different PR for it after this one is merged, will that be fine?

Reply:
Yes, doing that as a follow-up is perfectly fine 🙂